aiidalab_widgets_base/codes.py

```diff
@@ -71,35 +77,58 @@ def _get_codes(input_plugin):
                           with_computer='computer')
         results = querybuild.all()
 
-        # only codes on computers configured for the current user
+        # Only codes on computers configured for the current user.
         results = [r for r in results if r[0].is_user_configured(current_user)]
 
         codes = {"{}@{}".format(r[1].label, r[0].name): r[1] for r in results}
         return codes
 
-    def refresh(self, change=None):  # pylint: disable=unused-argument
-        """Refresh available codes."""
-        with self.output:
-            clear_output()
-            self.codes = self._get_codes(self.input_plugin)
-            options = list(self.codes.keys())
+    @staticmethod
+    def _get_full_code_name(code):
+        return "{}@{}".format(code.label, code.computer.name)
 
-            self.dropdown.options = options
+    def refresh(self, _=None):
+        """Refresh available codes.
 
-            if not options:
+        The job of this function is to look in AiiDA database, find available codes and
+        put them in the dropdown attribute."""
+
+        with self.output:
+            clear_output()
+            with self.hold_trait_notifications():
+                self.dropdown.options = self._get_codes(self.input_plugin)
+            if not self.dropdown.options:
                 print("No codes found for input plugin '{}'.".format(self.input_plugin))
                 self.dropdown.disabled = True
             else:
                 self.dropdown.disabled = False
 
-    @property
-    def selected_code(self):
-        """Returns a selected code."""
-        try:
-            return self.codes[self.dropdown.value]
-        except KeyError:
+    @validate('selected_code')
+    def _validate_selected_code(self, change):
+        """If code is provided, set it as it is. If code's name is provided,
+        select the code and set it."""
+        code = change['value']
+
+        # If code None, set value to None
+        if code is None:
             return None
 
+        # Check code by name.
+        if isinstance(code, str):
+            if code in self.codes:
+                return self.codes[code]
+            return None
```

Comment:
should there not be an exception raised in this case?
i.e. when setting it to a code that does not exist

Reply:
Agreed, will make it return an exception.